### PR TITLE
test for https://github.com/hazelcast/hazelcast/issues/13272

### PR DESCRIPTION
**Pull request by carstenartur**
add test for failing eviction since hazelcast 3.9.2

### CODE REVIEW

**Comment by carstenartur** [comment]
#13272 added another test with longer waittime not only for master but for 3.7.5 as well to see the difference in behavior.

**Comment by sancar** [comment]
@carstenartur Sorry for being late to look at this. 
There seems to be some commits mixed. I will ask you to send a clean commit if possible. 
@mmedenjak I am removing the client label and unassigning myself. This is not related to client. It seems the client commits are send by mistake.

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
@carstenartur many thanks for the reproducer. Can you please review & sign the [Hazelcast Contributor Agreement](https://hazelcast.atlassian.net/wiki/spaces/COM/pages/6357071/Hazelcast+Contributor+Agreement)?

I tried this on my local and the test failed (as expected) and I expect the same from the CI build. So in order to merge this while I work on a fix for #13272, you should `@Ignore("https://github.com/hazelcast/hazelcast/issues/13272")` those tests.

Update: in the meanwhile the CI finished with some checkstyle errors. While we are working on public access for the CI builds, here are the errors so you can take a look in the meanwhile:

```
12:04:32 /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java:0: File does not end with a newline.
12:04:32 /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.'.
12:04:32 /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java:43:1: File contains tab characters (this is the first instance).
```

you can also try checkstyle locally so you don't have to wait for CI results for that, just run maven with the `checkstyle` & `spotbugs` profiles enabled:

```
mvn -DskipTests -T1C -Pcheckstyle,spotbugs clean install
```

File: hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+	@Test
+	public void afterShortExpirationEntryShouldBeAway() throws InterruptedException {
+		final String KEY = "key";
+
+		Config hConfig = new Config().setInstanceName("instance")
+				.addMapConfig(new MapConfig().setName("test")
+						.setMaxSizeConfig(new MaxSizeConfig(200, MaxSizeConfig.MaxSizePolicy.FREE_HEAP_SIZE))
+						.setTimeToLiveSeconds(20));
+		final HazelcastInstance node = Hazelcast.newHazelcastInstance(hConfig);
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java]: In order for those tests to be able to run in parallel, create your `HazelcastInstance` as follows:
```
final HazelcastInstance node = createHazelcastInstance(hConfig);
```

Using `createHazelcastInstance` from test support class will create a `HazelcastInstance` that uses mock network, allowing parallel test execution without interference from other concurrent tests.

File: hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+	@Test
+	public void afterLongerExpirationEntryShouldBeAway() throws InterruptedException {
+		final String KEY = "key";
+
+		Config hConfig = new Config().setInstanceName("instance")
+				.addMapConfig(new MapConfig().setName("test")
+						.setMaxSizeConfig(new MaxSizeConfig(200, MaxSizeConfig.MaxSizePolicy.FREE_HEAP_SIZE))
+						.setTimeToLiveSeconds(20));
+		final HazelcastInstance node = Hazelcast.newHazelcastInstance(hConfig);
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java]: same here as above

Comment (carstenartur) [reply hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java]: Hi vbekiaris,
please check if it is looking now like you expected it.
Best regards,
Carsten

**Comment by vbekiaris** [comment]
@carstenartur your changes look good. Still checkstyle fails with:

```
/Users/vb/workspace/github/vbekiaris/hazelcast/hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java:22:8: Unused import - com.hazelcast.core.Hazelcast.
/Users/vb/workspace/github/vbekiaris/hazelcast/hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTimeoutTest.java:45:1: File contains tab characters (this is the first instance).
```

Please fix the unused import and make sure you format the file to use spaces instead of tabs to make checkstyle happy. You can verify locally that checkstyle passes with `mvn -Pcheckstyle -DskipTests clean install` so you don't have to wait for PR builder feedback.

**Comment by vbekiaris** [comment]
@carstenartur can you update the PR so checkstyle can pass and we can merge this PR? if you cannot make it in the following days and agree, I can integrate your commit in the fix PR #14867, fix the checkstyle issue and merge it along with the fix.

**Comment by carstenartur** [comment]
Hi vbekiaris,
please check if that is now what you need.
Best regards,
Carsten

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
PR builder failed at test `com.hazelcast.client.ClientClusterRestartEventTest.testMultiMemberRestart` (unrelated) (http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/1209)

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
@carstenartur thanks for the update, looks good!

**Comment by vbekiaris** [comment]
@carstenartur thanks for contributing the reproducer!

**Pull request merged**